cmd/login/clientcert.go

```diff
@@ -139,22 +157,100 @@ func fetchCredential(ctx context.Context, provider string, clientCertService cli
 	return secret, nil
 }
 
+func generateCredential(ctx context.Context, ca *corev1.Secret, config clientCertConfig) (*corev1.Secret, error) {
+	var caPEM, certPEM, keyPEM []byte
+	{
+		// Get the WCs CA data
+		caPEM = ca.Data["tls.crt"]
+		caCert, err := getCert(caPEM)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		caPrivKeyPEM := ca.Data["tls.key"]
+		caPrivKey, err := getPrivKey(caPrivKeyPEM)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		// Create the Certificate
+		exp, err := time.ParseDuration(config.ttl)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		serial, err := rand.Int(rand.Reader, new(big.Int).SetInt64(math.MaxInt64))
+		if err != nil {
+			return nil, microerror.Maskf(invalidConfigError, "Failed to generate client certificate serial number.")
+		}
+		certificate := &x509.Certificate{
+			SerialNumber:       serial,
+			SignatureAlgorithm: x509.SHA256WithRSA,
+			Subject: pkix.Name{
+				Organization: config.groups,
+				CommonName:   fmt.Sprintf("%s.%s.%s", serial.String(), config.clusterName, config.clusterBasePath),
+			},
+			NotBefore:   time.Now(),
+			NotAfter:    time.Now().Add(exp),
+			KeyUsage:    x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageKeyAgreement,
+			ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		}
+
+		// Create the key
+		certPrivKey, err := rsa.GenerateKey(rand.Reader, 2048)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		keyPEM = getPrivKeyPEM(certPrivKey)
+
+		// Sign the certificate
+		certBytes, err := x509.CreateCertificate(rand.Reader, certificate, caCert, &certPrivKey.PublicKey, caPrivKey)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		certPEM = getCertPEM(certBytes)
+
+	}
+	secret := &corev1.Secret{
+		Data: map[string][]byte{
```

Comment:
Should this also have `Type: "kubernetes.io/tls"` to be more standard? Not sure if Kubernetes enforces a specific structure on those kinds of secrets. See https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets